File: core/utils/result.ts

export type Result<T, E = Error> = 
  | { success: true; data: T }
  | { success: false; error: E };

export const Result = {
  ok: <T>(data: T): Result<T, never> => ({ success: true, data }),
  fail: <E>(error: E): Result<never, E> => ({ success: false, error }),
  
  // High-order utility
  tryCatch: async <T>(promise: Promise<T>): Promise<Result<T, Error>> => {
    try {
      const data = await promise;
      return Result.ok(data);
    } catch (error) {
      return Result.fail(error instanceof Error ? error : new Error(String(error)));
    }
  }
};

File: backend/LiliMagic.Api/Core/Result.cs
namespace LiliMagic.Api.Core;

public class Result<T>
{
    public bool Success { get; }
    public T? Data { get; }
    public string? Error { get; }
    public string? ErrorCode { get; }

    protected Result(bool success, T? data, string? error, string? errorCode)
    {
        Success = success;
        Data = data;
        Error = error;
        ErrorCode = errorCode;
    }

    public static Result<T> Ok(T data) => new(true, data, null, null);
    public static Result<T> Fail(string error, string? errorCode = null) => new(false, default, error, errorCode);
}
